src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/KeyConverters/ObjectKeyConverter.cs

```diff
@@ -0,0 +1,39 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+namespace System.Text.Json.Serialization.Converters
+{
+    internal sealed class ObjectKeyConverter : KeyConverter<object>
```

Comment:
Since we just use JsonDocument, how valuable is this?

What does GetHashCode() return from JsonDocument?

I think this scenario would be valuable once we add true polymorphic deserialization (and not just use JsonDocument).

Reply:
I don't quite understand the concern here. What this converter does is take the JSON property name as a string and create a `JsonElement` when the `TKey` is `object` in order to emulate the behavior of the value when `TValue` is `object`.

What does the `GetHashCode()` method have to do here?

Reply:
The string-based key creates a JsonElement which is used as a key in a dictionary, so GetHashCode() is called on the JsonElement by the dictionary during get\add operations.

Since JsonElement uses a default GetHashCode() implementation I don't see any value of having it as a key. If instead we supported polymorphic deserialization, the resulting object (not JsonElement) could have its own implementation of GetHashCode() and thus be useful as a key.

Reply:
I don't think we should not support such type, otherwise we would fail on round-tripping if someone have a property `Dictionary<object, object>`.

> What does GetHashCode() return from JsonDocument?

I think you mean `JsonElement`. It does return different values even for the same string so you should not have problems with get/add operations on the dictionary https://dotnetfiddle.net/pVlMi3

Reply:
We should wait until we have an [option to return boxed primitives over `JsonElement`](https://github.com/dotnet/runtime/issues/29960) before we consider allowing `typeof(object)` as a dictionary key.

Reply:
Why? We have object property support already, why not extend that for dictionary keys now with the same semantics?

Reply:
> We still need to flesh out the semantics/behavior when folks don't use that option, right?

That's fair.

For reasons called out in https://github.com/dotnet/runtime/pull/32909/files/53bfb9636b04b157b945ca59a1a6e958a791e00b#r385742410, calling `.ContainsKey` on the dictionary would only work if you "know" the specific `JsonElement` you are querying. The dictionary would only be useful when iterating over all key-value pairs. Lookup would be busted in general.